profiling/service/service.go

```diff
@@ -0,0 +1,125 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package service defines methods to register a gRPC client/service for a
+// profiling service that is exposed in the same server. This service can be
+// queried by a client to remotely manage the gRPC profiling behaviour of an
+// application.
+package service
+
+//go:generate protoc --go_out=plugins=grpc,paths=source_relative:../proto -I../proto service.proto
+
+import (
+	"context"
+	"errors"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal/profiling"
+	ppb "google.golang.org/grpc/profiling/proto"
+)
+
+// ProfilingConfig defines configuration options for the Init method.
+type ProfilingConfig struct {
+	// Setting this to true will enable profiling.
+	Enabled bool
+
+	// Profiling uses a circular buffer (ring buffer) to store statistics for
+	// only the last few RPCs so that profiling stats do not grow unbounded. This
+	// parameter defines the upper limit on the number of RPCs for which
+	// statistics should be stored at any given time. An average RPC requires
+	// approximately 2-3 KiB of memory for profiling-related statistics, so
+	// choose an appropriate number based on the amount of memory you can afford.
+	StreamStatsSize uint32
+
+	// To expose the profiling service and its methods, a *grpc.Server must be
+	// provided.
+	Server *grpc.Server
+}
+
+var errorNilServer = errors.New("profiling: no grpc.Server provided")
+
+// Init takes a *ProfilingConfig to initialize profiling (turned on/off
+// depending on the value set in pc.Enabled) and register the profiling service
+// in the server provided in pc.Server.
+func Init(pc *ProfilingConfig) error {
+	if pc.Server == nil {
+		return errorNilServer
+	}
+
+	if err := profiling.InitStats(pc.StreamStatsSize); err != nil {
+		return err
+	}
+
+	ppb.RegisterProfilingServer(pc.Server, &profilingServer{})
+
+	// Do this last after everything has been initialized and allocated.
+	profiling.Enable(pc.Enabled)
+
+	return nil
+}
+
+type profilingServer struct{}
+
+func (s *profilingServer) Enable(ctx context.Context, req *ppb.EnableRequest) (*ppb.EnableResponse, error) {
+	if req.Enabled {
+		grpclog.Infof("Enabling profiling")
+	} else {
+		grpclog.Infof("Disabling profiling")
+	}
+	profiling.Enable(req.Enabled)
```

Comment:
There are two ways to enable -- one through the service and another through `grpc/profiling.Enable`. The first one is required for manual enabling. The second is required for automatic enabling based on heuristics. 

If profiling is registered with multiple servers with different enable values, the final state will be that of whichever server was registered last. Init already protects against multiple calls using a CAS (see profiling.go in the internal package) -- all calls except the first one do nothing.

Could you elaborate a bit more on what a singleton `profilingServer` would do?

Reply:
There is only one "profiling thing" every binary, including enable/disable, and ring buffer.

Multiple servers point to the same profiling, and each can enable/disable, or even read from it independently. By singleton, I was suggesting we should have one server per binary, so it's 1:1.

And more than that, actually only one client can be support (at the same time?), because each `GetStat` RPC will read and delete all stats. If multiple clients are making this call, none of them will have the complete data.

Reply:
FWIW I find this API to be a bit awkward, but I don't have any suggestions to make it objectively better so I'm OK with it, especially as an initial commit for an experimental API.